tls/s2n_resume.c

```diff
@@ -769,187 +769,137 @@ struct s2n_ticket_key *s2n_find_ticket_key(struct s2n_config *config, const uint
+
+    /* Initialize Additional Authenticated Data */
     uint8_t aad_data[S2N_TICKET_AAD_LEN] = { 0 };
     struct s2n_blob aad_blob = { 0 };
-    POSIX_GUARD(s2n_blob_init(&aad_blob, aad_data, sizeof(aad_data)));
+    RESULT_GUARD_POSIX(s2n_blob_init(&aad_blob, aad_data, sizeof(aad_data)));
     struct s2n_stuffer aad = { 0 };
+    RESULT_GUARD_POSIX(s2n_stuffer_init(&aad, &aad_blob));
+    RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(&aad, key->implicit_aad, sizeof(key->implicit_aad)));
+    RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(&aad, key->key_name, sizeof(key->key_name)));
 
-    POSIX_GUARD(s2n_stuffer_read_bytes(from, key_name, s2n_array_len(key_name)));
-
-    key = s2n_find_ticket_key(conn->config, key_name);
-
-    /* Key has expired; do full handshake with New Session Ticket (NST) */
-    S2N_ERROR_IF(!key, S2N_ERR_KEY_USED_IN_SESSION_TICKET_NOT_FOUND);
-
-    POSIX_GUARD(s2n_stuffer_read(from, &iv));
-
-    POSIX_GUARD(s2n_blob_init(&aes_key_blob, key->aes_key, S2N_AES256_KEY_LEN));
-    POSIX_GUARD(s2n_session_key_alloc(&aes_ticket_key));
-    POSIX_GUARD_RESULT(s2n_aes256_gcm.init(&aes_ticket_key));
-    POSIX_GUARD_RESULT(s2n_aes256_gcm.set_decryption_key(&aes_ticket_key, &aes_key_blob));
-
-    POSIX_GUARD(s2n_stuffer_init(&aad, &aad_blob));
-    POSIX_GUARD(s2n_stuffer_write_bytes(&aad, key->implicit_aad, S2N_TICKET_AAD_IMPLICIT_LEN));
-    POSIX_GUARD(s2n_stuffer_write_bytes(&aad, key->key_name, S2N_TICKET_KEY_NAME_LEN));
-
+    /* Initialize blob to be decrypted */
     struct s2n_blob en_blob = { 0 };
     uint32_t en_blob_size = s2n_stuffer_data_available(from);
     uint8_t *en_blob_data = s2n_stuffer_raw_read(from, en_blob_size);
-    POSIX_ENSURE_REF(en_blob_data);
-    POSIX_GUARD(s2n_blob_init(&en_blob, en_blob_data, en_blob_size));
-    POSIX_GUARD(s2n_aes256_gcm.io.aead.decrypt(&aes_ticket_key, &iv, &aad_blob, &en_blob, &en_blob));
+    RESULT_ENSURE_REF(en_blob_data);
+    RESULT_GUARD_POSIX(s2n_blob_init(&en_blob, en_blob_data, en_blob_size));
 
+    RESULT_GUARD_POSIX(s2n_aes256_gcm.io.aead.decrypt(&aes_ticket_key, &iv, &aad_blob, &en_blob, &en_blob));
+
+    /* Parse decrypted state */
     struct s2n_blob state_blob = { 0 };
     uint32_t state_blob_size = en_blob_size - S2N_TLS_GCM_TAG_LEN;
-    POSIX_GUARD(s2n_blob_init(&state_blob, en_blob.data, state_blob_size));
+    RESULT_GUARD_POSIX(s2n_blob_init(&state_blob, en_blob.data, state_blob_size));
     struct s2n_stuffer state_stuffer = { 0 };
-    POSIX_GUARD(s2n_stuffer_init(&state_stuffer, &state_blob));
-    POSIX_GUARD(s2n_stuffer_skip_write(&state_stuffer, state_blob_size));
-    POSIX_GUARD_RESULT(s2n_deserialize_resumption_state(conn, &from->blob, &state_stuffer));
+    RESULT_GUARD_POSIX(s2n_stuffer_init(&state_stuffer, &state_blob));
+    RESULT_GUARD_POSIX(s2n_stuffer_skip_write(&state_stuffer, state_blob_size));
+
+    /* Session caching feature also uses this codepath */
```

Comment:
This probably needs a more detailed comment. Without knowing the history / digging into s2n_deserialize_resumption_state, branching on the arguments to s2n_deserialize_resumption_state looks pretty strange.

Should we also open an issue to fix that logic? It works, but it's pretty odd.

Reply:
What's the fix for this logic? No matter what, we're going to need a branch somewhere to make sure session cache isn't using the middle parameter.

Reply:
Somewhere, but here may not be the right place. It's very unclear why we're branching here, and passing "NULL" to the method is odd. I'd say ideally we should branch further down the stack where the parameter should either be used or not used. Like, where we're checking for NULL.

Reply:
I took out the branching logic. That middle parameter never gets used if you're entering the codepath from the decrypt function in TLS1.2.
I did however add a helper function so that the last code paragraph doesn't get executed if the ticket came from a session cache. 
Unfortunately theres still some code to parse the ticket key that is copied, but I think this is the most correct solution.

Reply:
Since there's confusion here, I was wondering if we could somehow write a test to prove no behavior change.

[s2n_decrypt_session_cache](https://github.com/search?q=repo%3Aaws%2Fs2n-tls+s2n_decrypt_session_cache&type=code) was only called by [s2n_resume_from_cache](https://github.com/aws/s2n-tls/blob/b20075ba20f911e4c78bc3d2b79a35682763dbb6/tls/s2n_resume.c#L465), which is only called [here](https://github.com/aws/s2n-tls/blob/b20075ba20f911e4c78bc3d2b79a35682763dbb6/tls/s2n_handshake_io.c#L1073). That means that s2n_decrypt_session_cache could only be called for a TLS1.2 server connection.

At the other end, the "ticket" parameter is only used by [s2n_deserialize_resumption_state](https://github.com/aws/s2n-tls/blob/b20075ba20f911e4c78bc3d2b79a35682763dbb6/tls/s2n_resume.c#L375-L385) for TLS1.2 tickets + client connections OR TLS1.3 tickets.

So I agree that the "ticket" parameter never gets used on the s2n_decrypt_session_cache path. We'd somehow need a TLS1.2 server decrypting a TLS1.3 ticket. Since s2n_resume_from_cache already assumes TLS1.2 tickets (it [only allocates enough space for a TLS1.2 ticket](https://github.com/aws/s2n-tls/blob/b20075ba20f911e4c78bc3d2b79a35682763dbb6/tls/s2n_resume.c#L450C17-L450C48)), we're not making that path worse.

But that also means that testing this would be hard. The best I can think of is to resume from the session cache and verify that 1) [conn->ticket](https://github.com/aws/s2n-tls/blob/b20075ba20f911e4c78bc3d2b79a35682763dbb6/tls/s2n_resume.c#L265-L275) isn't set 2) [no psk exists](https://github.com/aws/s2n-tls/blob/b20075ba20f911e4c78bc3d2b79a35682763dbb6/tls/s2n_resume.c#L296-L298). Not exactly a great test, so feel free not to implement it. Two people scrutinizing the code path is probably sufficient.